shell/src/main/java/alluxio/cli/fs/command/PersistCommand.java

```diff
@@ -55,63 +85,103 @@ public void validateArgs(CommandLine cl) throws InvalidArgumentException {
+      for (URIStatus s : statuses) {
+        queuePersist(s);
+      }
+    } else if (status.isPersisted()) {
+      System.out.println(status.getPath() + " is already persisted, skipping.");
+    } else {
+      mToPersist.add(uri);
+    }
   }
 
   @Override
   public int run(CommandLine cl) throws AlluxioException, IOException {
+    resetState();
+    int parallelism = 1;
+    if (cl.hasOption(PARALLELISM_OPTION.getLongOpt())) {
+      String parellismOption = cl.getOptionValue(PARALLELISM_OPTION.getLongOpt());
+      parallelism = Integer.parseInt(parellismOption);
+    }
     String[] args = cl.getArgs();
     for (String path : args) {
       AlluxioURI inputPath = new AlluxioURI(path);
       runWildCardCmd(inputPath, cl);
     }
-    return 0;
-  }
-
-  /**
-   * Persists a file or directory currently stored only in Alluxio to the UnderFileSystem.
-   *
-   * @param filePath the {@link AlluxioURI} path to persist to the UnderFileSystem
-   */
-  private void persist(AlluxioURI filePath) throws AlluxioException, IOException {
-    URIStatus status = mFileSystem.getStatus(filePath);
-    if (status.isFolder()) {
-      List<URIStatus> statuses = mFileSystem.listStatus(filePath);
-      List<String> errorMessages = new ArrayList<>();
-      for (URIStatus uriStatus : statuses) {
-        AlluxioURI newPath = new AlluxioURI(uriStatus.getPath());
-        try {
-          persist(newPath);
-        } catch (Exception e) {
-          errorMessages.add(e.getMessage());
-        }
-      }
-      if (errorMessages.size() != 0) {
-        throw new IOException(Joiner.on('\n').join(errorMessages));
-      }
-    } else if (status.isPersisted()) {
-      System.out.println(filePath + " is already persisted");
-    } else {
-      try {
-        FileSystemUtils.persistFile(mFileSystem, filePath);
-      } catch (InterruptedException e) {
-        Thread.currentThread().interrupt();
-        throw new RuntimeException(e);
-      } catch (TimeoutException e) {
-        throw new RuntimeException(e);
+    mTotalFilesToPersist = mToPersist.size();
+    System.out.println("Found " + mTotalFilesToPersist + " files to persist.");
+    ExecutorService service = Executors.newFixedThreadPool(parallelism);
+    List<Future<Void>> persistCalls = new ArrayList<>(parallelism);
+    for (int i = 0; i < parallelism; i++) {
+      persistCalls.add(service.submit(new PersistCallable()));
+    }
+    try {
+      for (Future<Void> call : persistCalls) {
+        call.get();
       }
-      System.out.println("persisted file " + filePath + " with size " + status.getLength());
+    } catch (ExecutionException e) {
+      System.out.println("Fatal error: " + e);
+    } catch (InterruptedException e) {
+      System.out.println("Persist interrupted, exiting.");
     }
+    return 0;
   }
 
   @Override
   public String getUsage() {
-    return "persist <path> [<path> ...]";
+    return "persist [--parallelism <#>] <path> [<path> ...]";
```

Comment:
Would it be easier to use `--parallel <#>` if `parallel` and `parallelism` do not have the obvious meaning difference

Reply:
Not strong feelings. Just lazy....Users like me always want the parameters to be shorter. If parallelism sounds more reasonable, then we better keep it.

Reply:
I've add a short option `-p`

Reply:
That's even better than `--parallel`, thanks!